File: backend/app/schemas/report.py
import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel


class CategoryCount(BaseModel):
    category: str
    count: int


class WeeklyCount(BaseModel):
    week: datetime
    count: int


class MonthlyRevenueDetail(BaseModel):
    month: datetime
    amount: Decimal
    auction_amount: Decimal
    direct_amount: Decimal
    sales_count: int


class RevenueOut(BaseModel):
    total_revenue: Decimal
    auction_revenue: Decimal
    direct_sales_revenue: Decimal
    sales_count: int
    monthly: list[MonthlyRevenueDetail]


class AuctionActivityOut(BaseModel):
    total: int
    upcoming: int
    live: int
    ended: int
    awarded: int
    total_bids: int
    weekly: list[WeeklyCount]


class DashboardOut(BaseModel):
    total_buyers: int
    total_sellers: int
    active_auctions: int
    total_listings: int
    published_listings: int
    sold_listings: int
    pending_approvals: int
    total_revenue: Decimal
    category_mix: list[CategoryCount]
    weekly_signups: list[WeeklyCount]


class SellerPerformance(BaseModel):
    seller_id: uuid.UUID
    seller_name: str
    listings: int
    sold: int
    avg_sale_price: Decimal
    success_rate: float


class BuyerActivity(BaseModel):
    buyer_id: uuid.UUID
    buyer_name: str
    bids: int
    won: int
    avg_spend: Decimal


class MarketingPerformanceOut(BaseModel):
    total: int
    sent: int
    scheduled: int
    draft: int


class BestSeller(BaseModel):
    property_id: uuid.UUID
    title: str
    final_price: Decimal
    bid_count: int


class AiInsightsOut(BaseModel):
    summary: str
    highlights: list[str]


class ConversionRatesOut(BaseModel):
    lead_to_buyer: float
    browse_to_bid: float
    bid_to_win: float

File: frontend/types/report.ts
export type ReportPropertyCategory = "residential" | "commercial";

export interface CategoryCount {
  category: ReportPropertyCategory;
  count: number;
}

export interface WeeklyCount {
  week: string;
  count: number;
}

export interface DashboardStats {
  total_buyers: number;
  total_sellers: number;
  active_auctions: number;
  total_listings: number;
  published_listings: number;
  sold_listings: number;
  pending_approvals: number;
  total_revenue: string;
  category_mix: CategoryCount[];
  weekly_signups: WeeklyCount[];
}

export interface MonthlyRevenueDetail {
  month: string;
  amount: string;
  auction_amount: string;
  direct_amount: string;
  sales_count: number;
}

export interface RevenueStats {
  total_revenue: string;
  auction_revenue: string;
  direct_sales_revenue: string;
  sales_count: number;
  monthly: MonthlyRevenueDetail[];
}

export interface AiInsights {
  total_revenue: string;
  avg_auction_value: string;
  top_category: string;
  buyer_count: number;
  seller_count: number;
  active_auctions: number;
  summary: string;
}

export interface BestSeller {
  property_id: string;
  title: string;
  category: string;
  total_bids: number;
  final_price: string | null;
}

export interface BuyerActivity {
  buyer_id: string;
  buyer_name: string;
  bids_placed: number;
  auctions_won: number;
  total_spent: string;
}

export interface ConversionRates {
  lead_to_buyer: number;
  browse_to_bid: number;
  bid_to_win: number;
}

export interface MarketingPerformance {
  total_campaigns: number;
  sent_campaigns: number;
  total_leads: number;
  converted_leads: number;
  conversion_rate: number;
}

export interface SellerPerformance {
  seller_id: string;
  seller_name: string;
  total_listings: number;
  sold_count: number;
  avg_sale_price: string;
  total_revenue: string;
}